### PR TITLE
CotE - Isawa Ujina

### LINKED ISSUE #2719: CotE - Isawa Ujina
image

### PR DESCRIPTION
**Pull request by jamesoddy**
Closes #2719 

Includes new RemoveFromGameAction

The new RemoveFromGameAction is a modified version of DiscardFromPlayAction with (particularly) a change to the destination.

Works ok with my tests, but there could be some interactions I have not considered.  Needs review.

RemoveFromGame is also expecting a card that is in play - that is probably not clear from the name of the action.  Any suggestions for a better action name?  Or a way to deal with out of play cards?

### CODE REVIEW

File: test/server/cards/06-CotE/IsawaUjina.spec.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                this.initiateConflict({
+                    attackers: [this.fushicho],
+                    defenders: [],
+                    ring: 'void'
+                });
+                this.noMoreActions();
+                this.player1.clickPrompt('No');
+                this.player1.clickCard(this.dojiHotaru);
+                expect(this.player1).toHavePrompt('Isawa Ujina');
+            });
```
Comment (jeremylarner) [inline test/server/cards/06-CotE/IsawaUjina.spec.js]: Might be worth a test making sure it interacts correctly with Kaede?

Comment (jamesoddy) [reply test/server/cards/06-CotE/IsawaUjina.spec.js]: Done.

File: server/game/GameActions/RemoveFromGameAction.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+export class RemoveFromGameAction extends CardGameAction {
+    name = 'removeFromGame';
+    eventName = EventNames.OnCardLeavesPlay;
+    cost = 'removing {0} from the game';
+    targetType = [CardTypes.Character, CardTypes.Attachment, CardTypes.Holding];
+
+    getEffectMessage(context: AbilityContext): [string, any[]] {
+        let properties = this.getProperties(context);
+        return['remove {0} from the game', [properties.target]];
+    }
```
[the remark this answers is not in the document]
Comment (jamesoddy) [reply server/game/GameActions/RemoveFromGameAction.ts]: Done.

File: server/game/GameActions/RemoveFromGameAction.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        return super.canAffect(card, context);
+    }
+
+    updateEvent(event, card: BaseCard, context: AbilityContext, additionalProperties): void {
+        additionalProperties.destination = Locations.RemovedFromGame;
+        this.updateLeavesPlayEvent(event, card, context, additionalProperties);
+    }
+
+    eventHandler(event): void {
+        event.destination = Locations.RemovedFromGame;
```
[the remark this answers is not in the document]
Comment (jamesoddy) [reply server/game/GameActions/RemoveFromGameAction.ts]: Looks like the destination was not being applied correctly in the `updateLeavesPlayEvent` function.  I have fixed that and it all seems ok now 😃 and this line has been removed.

**Comment by jamesoddy** [comment]
@jeremylarner I have made the requested changes and hopefully sorted the destination issue.  Please can you take another look and let me know if it all looks ok now.

**Pull request merged**